File: codegen/src/main/java/net/webpdf/codegen/extension/index/IndexEntry.java
package net.webpdf.codegen.extension.index;

import io.swagger.codegen.v3.CodegenModel;
import net.webpdf.codegen.extension.WebPDFExtension;

import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

@SuppressWarnings("unused")
public class IndexEntry implements Iterable<String> {

    private final List<String> exportedTypeNames = new ArrayList<>();
    private final String fileLocation;
    private final String packageLocation;
    private final CodegenModel model;

    public IndexEntry(String fileLocation, String packageLocation, CodegenModel model) {
        this.fileLocation = fileLocation;
        this.packageLocation = packageLocation;
        this.model = model;
    }

    public String getFileLocation() {
        return fileLocation;
    }

    public String getPackageLocation() {
        return packageLocation;
    }

    public CodegenModel getModel() {
        return model;
    }

    public String getExportedNames() {
        StringBuilder names = new StringBuilder();
        Iterator<String> itr = exportedTypeNames.iterator();
        while (itr.hasNext()) {
            names.append(itr.next());
            if (itr.hasNext()) {
                names.append(",");
            }
        }
        return names.toString();
    }

    public WebPDFExtension getWebPDFExtensions(String modelPackage) {
        return WebPDFExtension.determineExtension(model, modelPackage);
    }

    public List<String> getExportedTypeNames() {
        return exportedTypeNames;
    }

    public IndexEntry addExportedTypeName(String typeName) {
        this.exportedTypeNames.add(typeName);
        return this;
    }

    @Override
    public Iterator<String> iterator() {
        return exportedTypeNames.iterator();
    }

}

File: src/codegen/typescript/port/IndexEntry.ts
/**
 * Additional ordering metadata used for index generation.
 */
export type PortModelMeta = {
	parentClassName?: string;
	extendsName?: string;
};

/**
 * One generated index entry including exported type names and metadata.
 */
export class IndexEntry {
	private readonly exportedTypeNames: string[] = [];
	private readonly fileLocation: string;
	private readonly packageLocation: string;
	private readonly model: PortModelMeta;

	/**
	 * @param fileLocation Relative file location (without extension).
	 * @param packageLocation Package-style location.
	 * @param model Ordering metadata.
	 */
	public constructor(fileLocation: string, packageLocation: string, model: PortModelMeta) {
		this.fileLocation = fileLocation;
		this.packageLocation = packageLocation;
		this.model = model;
	}

	/**
	 * Returns the file location used in generated exports.
	 */
	public getFileLocation(): string {
		return this.fileLocation;
	}

	/**
	 * Returns the package-style location used for collision checks.
	 */
	public getPackageLocation(): string {
		return this.packageLocation;
	}

	/**
	 * Returns ordering metadata of the model.
	 */
	public getModel(): PortModelMeta {
		return this.model;
	}

	/**
	 * Returns exported names as a comma-separated string for `export { ... }`.
	 */
	public getExportedNames(): string {
		return this.exportedTypeNames.join(",");
	}

	/**
	 * Returns all exported type names for this entry.
	 */
	public getExportedTypeNames(): string[] {
		return this.exportedTypeNames;
	}

	/**
	 * Adds an exported type name to this entry.
	 *
	 * @param typeName Exported symbol name.
	 */
	public addExportedTypeName(typeName: string): IndexEntry {
		this.exportedTypeNames.push(typeName);
		return this;
	}
}
